src/PluginManager.js

```diff
@@ -26,30 +127,116 @@ class PluginManager {
+     */
+    getParserManager() {
+        return this.parserMgr;
+    }
+
+    /**
+     * Return the formatter manager for this plugin manager. This
+     * manages both the built-in formatters, and the formatters
+     * loaded from the plugins.
+     *
+     * @returns {FormatterManager} the formatter manager for this
+     * plugin manager.
+     */
+    getFormatterManager() {
+        return this.formatterMgr;
+    }
+
+    /**
+     * Return the rule manager for this plugin manager. This
+     * manages both the built-in rules, and the rules
+     * loaded from the plugins.
+     *
+     * @returns {RuleManager} the rule manager for this
+     * plugin manager.
+     */
+    getRuleManager() {
+        return this.ruleMgr;
+    }
+
+    /**
+     * Return the rules in this manager. This is from both the
+     * built-in rules and the rules loaded from the plugins.
+     *
+     * @returns {FormatterManager} the rule set for this plugin manager.
+     */
+    getRuleSet() {
+        return this.rules;
     }
 
     /**
-     * Load the named plugin.
+     * Add the already-loaded plugin to this manager.
      *
-     * @param {String} name plugin to load
-     * @returns {Promise} a promise to load the named plugin.
-     * @accept {Plugin} the loaded plugin
-     * @reject the plugin could not be found or loaded
+     * @param {Plugin} a plugin to add
      */
-    load(name) {
-        
+    add(plugin) {
+        if (!plugin) return;
+        this.parserMgr.add(plugin.getParsers());
+        this.formatterMgr.add(plugin.getFormatters());
+        this.ruleMgr.add(plugin.getRules());
+        this.ruleMgr.addRuleSetDefinitions(plugin.getRuleSets());
     }
 
     /**
-     * Return an array of handlers that handle the given path name based
-     * on things like the file name extension.
+     * Load the named plugin or plugins. If the names param is given
+     * as a string, a single plugin is loaded. If it is an array of strings,
+     * each named plugin is loaded. This method returns Promise
      *
-     * @param {String} pathName path to a file to match
-     * @param {String} type the type of plugin being sought
-     * @returns {Array.<Plugin>} an array of plugins that claim to handle
-     * the given path name
+     * @param {String|Array.<String>} names name or names of plugins to load
+     * @returns {Promise} a promise to load the named plugins.
+     * @accept {Array.<Object>} an array of promise statuses. The status
+     * field will either be "fulfilled" and the value field will be the
+     * Plugin instance, or "rejected" and the reason field will be filled
+     * with a description of why the plugin could not be loaded.
+     * @reject the plugins could not be found or loaded
      */
-    getHandlers(pathName, type) {
+    load(names) {
+        const API = getAPI();
+        if (typeof(name) === 'string') {
+            names = [ names ];
+        }
+        return Promise.allSettled(names.map(name => {
```

Comment:
Question) using Promise is the reason for using `import`?

Reply:
No, ES javascript does not allow require() any more in modules. The only way to do a dynamic import is through the import() keyword. (It's not really a function, but it looks like it.) The import keyword returns a promise. This is also the way things are imported the ilib-loader module for webpack, which is why it can only support async loading in webpack. To get around it in webpack, you can call ensureLocale() to pre-load everything for a locale asynchronously into the cache, and after it is done, you can call any ilib module synchronously for that locale and it will get all the data from the cache instead. The nodejs version of ilib-loader uses the "fs" library to load data files directly and synchronously, so it can support both sync and async loading for data. For ES modules, however, even nodejs only supports async loading with import(). Modules cannot be loaded synchronously ever.